Add support for prompt caching

kndl opened the pull request:
Add prompt caching to `TextContent` content type, and validate that the value gets serialized correctly for the Converse API as well as for the invoke model API with Claude and Nova LLMs.

# For consideration:
1. With these changes, only `TextContent` message content can request cache prompting. Doing so adds a cache point 
at that location in the conversation. [The API supports setting cache points on other content types](https://docs.aws.amazon.com/bedrock/latest/userguide/prompt-caching.html), but this PR does not.
1. Should we decide to make caching any type of content work as part of this PR, some light refactoring will likely be in order to avoid duplicating the cache point Pydantic models. Specifically, I personally like the design of the Converse API where the cache point is effectively a content type rather than an annotation on any content type. It might make sense to realign the e84-geoai-common API to work the same way.
1. Please note the serialization tests make sure the cache control format is accurate. It might be that you'd prefer to test that same logic differently.

# TODO
- [x] Code Todos
- [x] Validate Nova and Converse APIs work with live data.
- [x] Decide whether we should include include the API change described in list item 2 above as part of this PR.

kndl committed:
Add cache control to `TextContent`

Add cache control to `TextContent` and validate that the value gets
serialied correctly for the Converse API as well as for the invoke model
API with Claude and Nova LLMs.

element84-infra commented:
![](https://res.cloudinary.com/snyk/image/upload/r-d/scm-platform/snyk-pull-requests/pr-banner-default.svg)

### :tada: **Snyk checks have passed. No issues have been found so far.**

:white_check_mark: **security/snyk** check is complete. No issues have been found. [(View Details)](https://app.snyk.io/org/geoai/pr-checks/8fe4c5b7-7920-40de-9809-89f209198fed?refs=4c0f1d763952d8f7f08eea6d77f9870697574023&source=prChecksComment)

:white_check_mark: **license/snyk** check is complete. No issues have been found. [(View Details)](https://app.snyk.io/org/geoai/pr-checks/8fe4c5b7-7920-40de-9809-89f209198fed/license?refs=4c0f1d763952d8f7f08eea6d77f9870697574023&source=prChecksComment)

kndl commented on tests/llm/models/serialization/__init__.py:
These serialization tests felt like they made sense to validate the requests were being sent correctly. Let me know if there's a better way to test this.

kndl committed:
Nova: use `cachePoint` instead of `cache_point` in JSON payload

kndl committed:
Support Bedrock Converse cache read/write tokens in usage

kndl committed:
Validate that Claude works

kndl committed:
Add docstring to `_handle_content` for Converse API

Add a docstring explaining why the function returns a list of content
rather than a single content object.

kndl commented on tests/llm/models/test_claude.py:
```diff
@@ -254,3 +256,59 @@ def exec_image_gen_tool(tool_use_request: LLMToolUseContent) -> LLMToolResultCon
+    assert resp == expected_resp
+
+
+def test_large_system_prompt() -> None:
```
When run live, this test will succeed the first time, but it'll fail if it's run again in the next 5 minutes.

Because these tests aren't run live particularly often, that might be fine. I only use the live tests to inspect the actual results.

A couple strategies we could use for repeatability:
* Gen a UUID and include it in the message. This will ensure that we always write to the cache.
* Assert that the cache is read from if it is not written to.

kndl committed:
Make library's API mirror the Converse API

After discussing, we determined that we preferred the Bedrock Converse's
API style of treating a cache point as its own content type rather than
the Nova/Claude invoke model-style API of making the caching part of
each individual content type.

Update the API the library exposes to mirror the Bedrock style, and
update the transformation for the 3 APIs we use to support the new API.

kndl committed:
Add a base class for `ClaudeCacheableContent`

Reduce boilerplate declaring the caching fields in the Claude content
types by introducing this base type.

kndl committed:
Add `NovaCacheableContent`

This additional type reduces boilerplate and matches the type pattern
that exists with the Claude API.

kndl committed:
Claude `should_request_prompt_cache` -> `should_cache`

Simpler name. More alignment with Nova API.

AdeelH commented on src/e84_geoai_common/llm/models/claude.py:
```diff
@@ -53,12 +53,12 @@ class ClaudeCacheControl(BaseModel):
     def cache_control(self) -> ClaudeCacheControl | None:
-        if self.should_request_prompt_cache:
+        if self.should_cache:
             return ClaudeCacheControl()
```
Non-blocking: it's okay to have `cache_control` as a direct (instead of computed) field. The main goal of these model classes is to closely align with the inputs/outputs of the Claude/Nova APIs. Being user-friendly is less important, especially because users are expected to use the general models in `llm.py` rather than these ones directly.

AdeelH commented on src/e84_geoai_common/llm/core/llm.py:
```diff
@@ -57,7 +55,17 @@ class LLMToolResultContent(BaseModel):
 
-LLMMessageContentType = TextContent | Base64ImageContent | LLMToolUseContent | LLMToolResultContent
+class CachePointContent(BaseModel):
+    """Tool invocation request."""
```
Docstring

kndl committed:
Do not use computed fields in Claude/Nova models

As pointed out in code review, these models are intended to be transit
types. They should prefer making it easy to reason about the JSON they
generate rather than making it ergonomic to instantiate for a developer.

kndl committed:
Fix docstring
